packages/mui-system/src/cssVars/useCurrentColorScheme.ts

```diff
@@ -278,35 +283,41 @@ export default function useCurrentColorScheme<SupportedColorScheme extends strin
     // Intentionally use deprecated listener methods to support iOS & old browsers
     media.addListener(handler);
     handler(media);
-
-    return () => media.removeListener(handler);
+    return () => {
+      media.removeListener(handler);
+    };
   }, []);
 
   // Handle when localStorage has changed
   React.useEffect(() => {
-    const handleStorage = (event: StorageEvent) => {
-      const value = event.newValue;
-      if (
-        typeof event.key === 'string' &&
-        event.key.startsWith(colorSchemeStorageKey) &&
-        (!value || joinedColorSchemes.match(value))
-      ) {
-        // If the key is deleted, value will be null then reset color scheme to the default one.
-        if (event.key.endsWith('light')) {
-          setColorScheme({ light: value as SupportedColorScheme | null });
+    if (storageWindow) {
```

Comment:
Is this operating on the same key as `useLocalStorageState`? Can't it just use the hook instead? I don't think the storage value should be manipulated by anything else, it will definitely lead to bugs.

Reply:
It does, but the TODO I have left is to move to a point where the mode is to be handled by `eseCurrentColorScheme` and not `useLocalStorageState`.

Reply:
The way I envisioned this was (pseudocode)

```tsx
// have these hooks available in the docs
function usePaletteModeUserPreference() {
  const [mode, setMode] = useLocalStorageState('colorScheme', 'system');
  return { mode, setMode };
}

function usePaletteMode() {
  const { mode: userMode } = usePaletteModeUserPreference();
  const systemColor = useSystemColor();
  return userMode === 'system' ? systemColor : userMode;
}


// for theme provider
const paletteMode = usePaletteMode()
const theme = paletteMode === 'dark' ? darkTheme : lightTheme;

// anywhere for theme switchers
const { mode, setMode } = usePaletteModeUserPreference();
<Button active={mode === 'dark'} onClick={() => setMode('dark')}>dark</Button>
<Button active={mode === 'light'} onClick={() => setMode('light')}>light</Button>
```

The nice thing about `useLocalStorage` is that it acts as global state, so you don't ever need to add a shared state variable somewhere and create a context for it. Just the hook works anywhere you use it. This system worked very well for the Toolpad app theme switcher. I might be oversimplifying for the docs, I'd have to look deeper into it, there may be blockers to adopt the same way of doing a theme switcher

Reply:
@Janpot Yes but what I see is that the mode is a lot more integrated with MUI System once you use CSS Variables and zero runtime than it was with emotion (what Toolpad uses).
MUI System almost needs to completely own that logic. If it's possible to separate things then I'm 💯 for it. I had this feeling that hard to understand how things work because of how closely things are linked to gather @siriwatknp.